wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java

```diff
@@ -485,4 +492,84 @@ public String toString(E value) {
       this.value = value;
     }
   }
+
+  private static final class MapProtoAdapter<K, V> extends ProtoAdapter<Map<K, V>> {
+    private final MapEntryProtoAdapter<K, V> entryAdapter;
+
+    MapProtoAdapter(ProtoAdapter<K> keyAdapter, ProtoAdapter<V> valueAdapter) {
+      super(FieldEncoding.LENGTH_DELIMITED, null);
+      entryAdapter = new MapEntryProtoAdapter<>(keyAdapter, valueAdapter);
+    }
+
+    @Override public int encodedSize(Map<K, V> value) {
+      throw new UnsupportedOperationException("Repeated values can only be sized with a tag.");
+    }
+
+    @Override public int encodedSizeWithTag(int tag, Map<K, V> value) {
+      int size = 0;
+      for (Map.Entry<K, V> entry : value.entrySet()) {
+        size += entryAdapter.encodedSizeWithTag(tag, entry);
+      }
+      return size;
+    }
+
+    @Override public void encode(ProtoWriter writer, Map<K, V> value) {
+      throw new UnsupportedOperationException("Repeated values can only be encoded with a tag.");
+    }
+
+    @Override public void encodeWithTag(ProtoWriter writer, int tag, Map<K, V> value)
+        throws IOException {
+      for (Map.Entry<K, V> entry : value.entrySet()) {
+        entryAdapter.encodeWithTag(writer, tag, entry);
+      }
+    }
+
+    @Override public Map<K, V> decode(ProtoReader reader) throws IOException {
+      K key = null;
+      V value = null;
+
+      long token = reader.beginMessage();
+      for (int tag; (tag = reader.nextTag()) != -1;) {
+        switch (tag) {
+          case 1: key = entryAdapter.keyAdapter.decode(reader); break;
+          case 2: value = entryAdapter.valueAdapter.decode(reader); break;
+          default: // Ignore unknown tags in map entries.
+        }
+      }
+      reader.endMessage(token);
+
+      if (key == null) throw new IllegalStateException("Map entry with null key");
+      if (value == null) throw new IllegalStateException("Map entry with null value");
+      return Collections.singletonMap(key, value);
+    }
+
+    @Override public Map<K, V> redact(Map<K, V> value) {
+      return Collections.emptyMap();
```

Comment:
Yep. You either mark the field as redacted and get empty or we ask the values to redact themselves.